File: python/mc_bow_agent/ballistic.py
"""Tick-accurate Minecraft arrow simulator + ballistic pitch solver.

Reference oracle for the M0 physics-parity gate. Per-tick update order
(minecraft.wiki/Arrow): velocity *= drag; velocity.y -= gravity; position += velocity.
VERIFY this order against the live game in M0 before trusting it for labels.
"""
import math
import numpy as np

from . import constants as C


def look_vector(yaw_deg, pitch_deg):
    """Minecraft look unit vector for (yaw, pitch) in degrees (pitch down-positive)."""
    yaw = math.radians(yaw_deg)
    pitch = math.radians(pitch_deg)
    return np.array([
        -math.sin(yaw) * math.cos(pitch),
        -math.sin(pitch),
        math.cos(yaw) * math.cos(pitch),
    ])


def simulate_arrow(speed, pitch_deg, yaw_deg=0.0, start=(0.0, 0.0, 0.0),
                   max_ticks=400, drag=C.ARROW_DRAG, gravity=C.ARROW_GRAVITY):
    """Return an (max_ticks+1, 3) array of arrow positions, starting at `start`."""
    pos = np.array(start, dtype=np.float64)
    vel = look_vector(yaw_deg, pitch_deg) * speed
    traj = [pos.copy()]
    for _ in range(max_ticks):
        vel = vel * drag
        vel[1] -= gravity
        pos = pos + vel
        traj.append(pos.copy())
    return np.array(traj)


def height_at_distance(speed, pitch_deg, horiz_dist, **kw):
    """Interpolated y when the arrow first reaches horiz_dist in the xz-plane.
    Returns None if it never gets that far."""
    traj = simulate_arrow(speed, pitch_deg, **kw)
    prev = traj[0]
    for cur in traj[1:]:
        hz_prev = math.hypot(prev[0], prev[2])
        hz_cur = math.hypot(cur[0], cur[2])
        if hz_cur >= horiz_dist:
            if hz_cur == hz_prev:
                return float(cur[1])
            t = (horiz_dist - hz_prev) / (hz_cur - hz_prev)
            return float(prev[1] + t * (cur[1] - prev[1]))
        prev = cur
    return None


def solve_pitch(speed, horiz_dist, height_delta, lo=-45.0, hi=30.0, iters=60):
    """Pitch (deg, down-positive) so a freshly-loosed arrow passes through
    (horiz_dist, height_delta) -- the lower (direct) arc. Bisection on a
    function that is monotonically decreasing in pitch."""
    def f(p):
        h = height_at_distance(speed, p, horiz_dist)
        return None if h is None else h - height_delta

    for _ in range(iters):
        mid = (lo + hi) / 2.0
        fm = f(mid)
        if fm is None or fm > 0:   # aimed too far up (or fell short) -> bigger pitch
            lo = mid
        else:
            hi = mid
    return (lo + hi) / 2.0


def lead_target(speed, shooter_eye, mob_pos, mob_vel, aim_offset=0.0, iters=5):
    """Fixed-point lead: predict where the arrow intercepts a constant-velocity
    mob. Returns the predicted aim point (x, y, z). aim_offset raises the point
    to the mob hitbox centre. mob_vel is blocks/tick."""
    target = np.array(mob_pos, dtype=np.float64) + np.array([0.0, aim_offset, 0.0])
    eye = np.array(shooter_eye, dtype=np.float64)
    vel = np.array(mob_vel, dtype=np.float64)
    for _ in range(iters):
        horiz = math.hypot(target[0] - eye[0], target[2] - eye[2])
        # crude time-of-flight estimate using the drag-decayed mean horizontal speed
        tof = horiz / max(speed * 0.85, 1e-6)
        target = np.array(mob_pos, dtype=np.float64) + vel * tof + np.array([0.0, aim_offset, 0.0])
    return tuple(target)

File: mod/src/main/java/com/mcbowagent/oracle/Ballistic.java
package com.mcbowagent.oracle;

import com.mcbowagent.config.RecorderConfig;

/**
 * Tick-accurate Minecraft arrow simulator + ballistic pitch solver. Java port of
 * {@code python/mc_bow_agent/ballistic.py} — kept in sync so the oracle (privileged
 * aimbot, F10) and the vision runtime (Python) solve the same drop with the same
 * answer. Update order per tick (minecraft.wiki/Arrow): v *= drag; v.y -= gravity;
 * pos += v.
 *
 * Coordinate convention here mirrors the Python module:
 *   pitch DOWN is POSITIVE (Minecraft convention).
 *   yaw 0 looks along +Z; we always solve in the player-target horizontal plane
 *   so yaw drops out of the bisection (we model XZ as a single "horiz" axis).
 */
public final class Ballistic {

    private Ballistic() {}

    /**
     * Returns the arrow's y at the moment it first reaches horizDist in the XZ plane
     * (linearly interpolated between the bracketing ticks). NaN if it never reaches
     * that distance within maxTicks — caller treats that as "out of range".
     */
    public static double heightAtDistance(double speed, double pitchDeg, double horizDist, int maxTicks) {
        double pitchRad = Math.toRadians(pitchDeg);
        // initial velocity from look direction. Pitch DOWN is positive -> y-component is -sin(pitch)
        double cosP = Math.cos(pitchRad);
        double vx = 0.0;                    // we collapse XZ to one axis: vz is the radial speed
        double vy = -Math.sin(pitchRad) * speed;
        double vz = cosP * speed;
        double px = 0.0, py = 0.0, pz = 0.0;
        double prevHz = 0.0, prevY = 0.0;
        for (int t = 0; t < maxTicks; t++) {
            // tick update: drag, gravity, then move
            vx *= RecorderConfig.ARROW_DRAG;
            vy *= RecorderConfig.ARROW_DRAG;
            vz *= RecorderConfig.ARROW_DRAG;
            vy -= RecorderConfig.ARROW_GRAVITY;
            px += vx;
            py += vy;
            pz += vz;
            double hz = Math.hypot(px, pz);
            if (hz >= horizDist) {
                if (hz == prevHz) return py;
                double f = (horizDist - prevHz) / (hz - prevHz);
                return prevY + f * (py - prevY);
            }
            prevHz = hz;
            prevY = py;
        }
        return Double.NaN;
    }

    /**
     * Solve the pitch (deg, down-positive) that lands a freshly-loosed arrow at
     * (horizDist, heightDelta). Bisection on the lower (direct) arc — the lookup
     * function is monotonically decreasing in pitch. Returns the bracket midpoint
     * after `iters` halvings (~1e-13 deg precision in 60 iters; 20 is plenty).
     */
    public static double solvePitch(double speed, double horizDist, double heightDelta) {
        double lo = -45.0;   // very steep up
        double hi = 30.0;    // moderate down
        for (int i = 0; i < 40; i++) {
            double mid = (lo + hi) / 2.0;
            double y = heightAtDistance(speed, mid, horizDist, 400);
            // NaN -> aimed too far up, never reached horizDist -> need bigger pitch (more down)
            if (Double.isNaN(y) || y - heightDelta > 0) {
                lo = mid;
            } else {
                hi = mid;
            }
        }
        return (lo + hi) / 2.0;
    }
}
